### ISSUE TITLE
[feat] geotoolz.matched_filter — composable hyperspectral matched filter (CH₄/CO₂)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/geotoolz#31]
> **Suggested title:** `[feat] geotoolz.matched_filter` — composable hyperspectral matched filter (CH₄/CO₂)
> **Suggested labels:** `enhancement`, `module:matched_filter`, `priority:high`, `domain:methane`

## User story

As a methane / CO₂ researcher applying a matched filter to hyperspectral imagery (EMIT, PRISMA, EnMAP, AVIRIS-NG, GHGSat), I want every piece of the MF pipeline — background statistics, covariance estimation, target signature, per-pixel scoring, SNR, detection threshold — exposed as composable Operators on `GeoTensor`, so that the whole retrieval is one declarative graph from radiance cube to enhancement map to detection mask.

## Motivation

A full-featured matched filter already exists in `jejjohnson/research_notebook` at `projects/plume_simulation/src/plume_simulation/matched_filter/`. It’s a six-module subsystem with a deliberate split:

- **Estimation** — `background.py`, `cluster.py`, `streaming.py`. NumPy + scikit-learn. Not on any autodiff path — the MF treats the background as a fixed input.
- **Application** — `core.py`. Pure JAX. Per-pixel `(x − μ)ᵀ Σ⁻¹ t / (tᵀ Σ⁻¹ t)` evaluated as one einsum across the cube.
- **Target signature** — `target.py`. Built from an existing JAX `RadianceObservationModel` via `jax.jvp` (tangent-linear) or finite-amplitude secant. No duplicate Beer–Lambert code.
- **Covariance dispatch** — every estimator returns a `lineax.AbstractLinearOperator` (dense / shrunk / low-rank+Tikhonov via `gaussx.LowRankUpdate`), and the kernel calls `gaussx.solve`, so Cholesky / Woodbury / diagonal paths are transparent to the caller.
- **I/O** — `io.py`. xarray wrappers for multi-scene workflows.

The package docstring also notes the **one-CG-iteration limit of dual / PSAS 3D-Var** with a flat prior — this is the right module for users who later want to graduate to a full variational retrieval.

This issue wraps the reference implementation in a `geotoolz` Operator surface. It does *not* propose reimplementing the math — that work is done, tested, and uses the right libraries.

## Architectural decision (please review before implementation)

The reference has heavy deps: `jax`, `lineax`, `gaussx`, `scikit-learn`. `geotoolz` operates on `georeader.GeoTensor` (numpy subclass). Three integration paths:

**A. Thin wrappers + optional extra.** Keep the implementation in `plumax` (or extract to a `methane-mf` package), import conditionally under `geotoolz.matched_filter`. Install via `pip install geotoolz[matched_filter]`.

**B. Vendor into geotoolz.** Lists `jax`/`lineax`/`gaussx` as required deps. Largest install footprint.

**C. Sister package `geotoolz-mf`.** Plugin-style, separate repo.

**Recommendation: A.** The MF is substantial enough to deserve its own home, but tight enough to the methane workflow that it should be adjacent to the rest of the operator surface, not a sibling repo. `plumax` is the natural source of truth given the existing `RadianceObservationModel` lives there.

A second decision to surface in this issue: the reference is **functional** (`apply_image`, `estimate_cov_*`); `geotoolz` is **OO-Operator**. The plan below adapts function → Operator one-to-one. State (estimated `(μ, Σ_op, t)`) lives on the Operator instance so it survives YAML/Hydra-zen serialisation.

## Proposed API

```python
# ── Application kernel ──────────────────────────────────────────────────────
class MatchedFilter(Operator):
    """Per-pixel matched filter score map.
    Score: y = (x − μ)ᵀ Σ⁻¹ t / (tᵀ Σ⁻¹ t).
    Wraps `matched_filter.core.apply_image`.
    """
    mean: np.ndarray | None = None             # provided, or set by fit-on-call
    cov_op: LinearOperator | None = None       # lineax operator
    target: np.ndarray | None = None
    fit_on_call: bool = False                  # estimate (μ, Σ) from the input cube
    mean_method: str = "median"                # used if fit_on_call
    cov_method: str = "ledoit_wolf"            # used if fit_on_call

class MatchedFilterPixel(Operator):
    """Single-pixel score — wraps `apply_pixel`. Useful for unit tests / interactive."""
    mean: np.ndarray
    cov_op: LinearOperator
    target: np.ndarray

# ── Statistics ──────────────────────────────────────────────────────────────
class MatchedFilterSNR(Operator):
    """α · √(tᵀ Σ⁻¹ t) — theoretical detection SNR at amplitude α."""
    amplitude: float
    cov_op: LinearOperator
    target: np.ndarray

class DetectionThreshold(Operator):
    """Score threshold for a Gaussian false-alarm rate: Φ⁻¹(1 − FAR) / √(tᵀΣ⁻¹t)."""
    false_alarm_rate: float
    cov_op: LinearOperator
    target: np.ndarray

class ValidateMFInputs(Operator):
    """Fail-fast pass-through: raises a clear ValueError on degenerate (Σ, t)."""
    cov_op: LinearOperator
    target: np.ndarray

# ── Mean estimators (numpy / sklearn) ───────────────────────────────────────
class EstimateMean(Operator):
    method: str = "mean"  # "mean" | "median" | "trimmed" | "huber"
    trim_proportion: float = 0.1
    huber_c: float = 1.345

# ── Covariance estimators (return lineax operators) ─────────────────────────
class EstimateCovEmpirical(Operator):
    mean: np.ndarray | None = None
    ridge: float = 0.0

class EstimateCovShrunk(Operator):
    mean: np.ndarray | None = None
    method: str = "ledoit_wolf"  # "ledoit_wolf" | "oas"

class EstimateCovLowRank(Operator):
    mean: np.ndarray | None = None
    rank: int = 10
    tikhonov: float = 1e-3
    random_state: int | None = 0
    n_oversamples: int = 10

# ── Heterogeneous backgrounds ───────────────────────────────────────────────
class GMMClusterBackground(Operator):
    n_clusters: int
    cov_estimator: str = "ledoit_wolf"
    random_state: int | None = 0
    bayesian: bool = False  # BayesianGaussianMixture prunes unused components

class AdaptiveWindowBackground(Operator):
    """Per-pixel local (μ, diag(σ²)) over a square window."""
    window_size: int = 7
    pad_mode: str = "reflect"

class ApplyClusterMF(Operator):
    """Per-pixel MF dispatched by cluster label."""
    cluster: ClusterBackground
    target: np.ndarray

# ── Streaming / multi-scene ─────────────────────────────────────────────────
class StreamingBackground(Operator):
    """Aggregate (μ, Σ) across an iterable of cubes via parallel Welford."""
    cubes: Iterable[GeoTensor]
    cov_kind: str = "shrunk"  # "empirical" | "shrunk"

# The WelfordAccumulator stays a plain dataclass — re-exported from this module
# for users who want fine-grained control. Not wrapped as an Operator because
# its semantics (update / merge / finalise) don't fit the call-once model.

# ── Target signature (JAX forward model) ────────────────────────────────────
class LinearTargetFromObs(Operator):
    """t = jvp(H, x_b, δx) — tangent-linear at the background state."""
    obs_model: RadianceObservationModel
    vmr_background: np.ndarray | None = None
    pattern: str | np.ndarray = "uniform"  # "uniform" | "impulse" | array
    pixel: tuple[int, int] | None = None
    linear_forward: bool = False  # use 1 − Δτ instead of exp(−Δτ)

class NonlinearTargetFromObs(Operator):
    """t = H(x_b + α·δx) − H(x_b) — Beer–Lambert saturation for strong plumes."""
    obs_model: RadianceObservationModel
    vmr_background: np.ndarray | None = None
    amplitude: float = 1.0
    pattern: str | np.ndarray = "uniform"
    pixel: tuple[int, int] | None = None

# ── Convenience composites ──────────────────────────────────────────────────
class ColumnEnhancement(Operator):
    """EstimateMean → EstimateCovShrunk → LinearTargetFromObs → MatchedFilter."""
    gas: str = "CH4"                                # for the absorption template
    sensor: str = "EMIT"                            # for the spectral grid
    obs_model: RadianceObservationModel | None = None
    mean_method: str = "median"
    cov_method: str = "ledoit_wolf"                 # or "lowrank"
    target_pattern: str = "uniform"
    rank: int | None = None                         # required if cov_method="lowrank"
    tikhonov: float | None = None
```

## Demo use cases

```python

import geotoolz as gz
import numpy as np

# ── 1. Imperative graph: estimate (μ, Σ, t) from one cube, apply MF to it ──
# Each estimator reads the same cube; MatchedFilter then combines their
# outputs with that cube. This is intentionally NOT a Sequential — three of
# the four steps return non-GeoTensor outputs (vector / lineax op / vector)
# that are inputs to MatchedFilter rather than pipeline stages.
cube = gz.Sequential([
    gz.io.ReadBounds(src=emit_l1b, bounds=aoi),
    gz.qa.MaskNoData(),
])()

mu       = gz.matched_filter.EstimateMean(method="median")(cube)
Sigma_op = gz.matched_filter.EstimateCovShrunk(method="ledoit_wolf", mean=mu)(cube)
t        = gz.matched_filter.LinearTargetFromObs(
               obs_model=emit_radiance_model, pattern="uniform",
           )(cube)
gz.matched_filter.ValidateMFInputs(cov_op=Sigma_op, target=t)()    # fail-fast

enhancement = gz.matched_filter.MatchedFilter(
    mean=mu, cov_op=Sigma_op, target=t,
)(cube)

# ── 2. Convenience: one Operator wraps the same graph with sensor defaults ─
enhancement = gz.matched_filter.ColumnEnhancement(
    gas="CH4", sensor="EMIT", obs_model=emit_radiance_model,
    mean_method="median", cov_method="ledoit_wolf",
)(cube)

# ── 3. Patching: spatially-adaptive background via per-tile (μ, Σ) ─────────
# Heterogeneous scenes (urban/rural, land/water) smear together statistically
# distinct populations under a single global background. Estimate (μ, Σ) per
# tile, share one target across all tiles, stitch the score map back.
t = gz.matched_filter.LinearTargetFromObs(obs_model=emit_obs)(cube)

def mf_per_tile(tile):
    mu_t       = gz.matched_filter.EstimateMean(method="median")(tile)
    Sigma_t_op = gz.matched_filter.EstimateCovShrunk(method="oas", mean=mu_t)(tile)
    return gz.matched_filter.MatchedFilter(
        mean=mu_t, cov_op=Sigma_t_op, target=t,
    )(tile)

tiles  = gz.patches.ExtractPatches(size=(256, 256), stride=(192, 192))(cube)
scores = [mf_per_tile(tile) for tile in tiles]
enhancement = gz.patches.StitchPatches(
    target_shape=cube.shape[:2], target_transform=cube.transform,
    target_crs=cube.crs, blend="feather", feather_width=24,
)(scores)

# ── 4. Streaming: Welford-aggregated background across a flight campaign ──
# Per-scene statistics are noisy when clear-air pixels are scarce; accumulate
# (μ, Σ) across many scenes via parallel Welford (merges chunks exactly),
# then re-use that fixed background for every scene.
bg = gz.matched_filter.StreamingBackground(
    cubes=(gz.io.ReadWindow(src=p, window=full)() for p in campaign_scenes),
    cov_kind="shrunk",
)()
mu_global, Sigma_global_op = bg.mean, bg.cov_op

ref_cube = gz.io.ReadWindow(src=campaign_scenes[0], window=full)()
t_global = gz.matched_filter.LinearTargetFromObs(obs_model=emit_obs)(ref_cube)
mf = gz.matched_filter.MatchedFilter(
    mean=mu_global, cov_op=Sigma_global_op, target=t_global,
)
enhancements = {p: mf(gz.io.ReadWindow(src=p, window=full)())
                for p in campaign_scenes}

# ── 5. Graph: GMM-clustered background for heterogeneous scenes ────────────
# Per-pixel cluster assignment → each cluster gets its own (μ_k, Σ_k) → MF
# dispatches the right local stats per pixel.
cluster_bg = gz.matched_filter.GMMClusterBackground(
    n_clusters=4, cov_estimator="ledoit_wolf", random_state=0,
)(cube)
t = gz.matched_filter.LinearTargetFromObs(obs_model=emit_obs)(cube)
enhancement = gz.matched_filter.ApplyClusterMF(
    cluster=cluster_bg, target=t,
)(cube)

# ── 6. Multi-modal: lock (Σ, t, τ) on a reference scene, apply to others ───
# Use one clear-air reference to estimate covariance, target, and analytic
# Gaussian detection threshold; apply MF (with a freshly-estimated μ per
# scene) and that fixed threshold to every target scene.
ref = gz.io.ReadWindow(src=clear_air_emit, window=full)()
Sigma_ref = gz.matched_filter.EstimateCovShrunk()(ref)
t_ref     = gz.matched_filter.LinearTargetFromObs(obs_model=emit_obs)(ref)
tau       = gz.matched_filter.DetectionThreshold(
    false_alarm_rate=1e-6, cov_op=Sigma_ref, target=t_ref,
)()
snr_1ppm  = gz.matched_filter.MatchedFilterSNR(
    amplitude=1.0, cov_op=Sigma_ref, target=t_ref,
)()

for scene in target_scenes:
    mu  = gz.matched_filter.EstimateMean(method="median")(scene)
    enh = gz.matched_filter.MatchedFilter(
        mean=mu, cov_op=Sigma_ref, target=t_ref,
    )(scene)
    detections = gz.plume.PlumeMask(threshold=float(tau))(enh)

```

## Plan

- [ ] Decide integration path A / B / C above — file as a discussion before any code
- [ ] If A: pin the source-of-truth (`plumax`? a new `methane-mf`? `research_notebook` for now?) and add it as an optional install extra
- [ ] Define a minimal `GeoTensor ↔ jax.Array` bridge: strip-and-restore geo metadata around the JAX kernel
- [ ] Decide Operator state semantics: `fit_on_call=True` populates `(mean, cov_op, target)` from the input; pre-supplied values skip the fit (essential for multi-scene re-use)
- [ ] Wrap `apply_image`, `apply_pixel` → `MatchedFilter`, `MatchedFilterPixel`
- [ ] Wrap `matched_filter_snr`, `detection_threshold`, `validate_mf_inputs`
- [ ] Wrap `estimate_mean` with all four methods (mean / median / trimmed / huber)
- [ ] Wrap `estimate_cov_empirical`, `estimate_cov_shrunk`, `estimate_cov_lowrank`
- [ ] Wrap `gmm_cluster_background`, `adaptive_window_background`
- [ ] Implement `ApplyClusterMF` — per-pixel dispatch on cluster labels using the same `gaussx.solve` kernel
- [ ] Re-export `WelfordAccumulator` and wrap `streaming_background`
- [ ] Wrap `linear_target_from_obs`, `nonlinear_target_from_obs`
- [ ] `ColumnEnhancement` composite with sensor + gas presets (HITRAN-derived templates owned by the source-of-truth package, not by `geotoolz`)
- [ ] xarray / GeoTensor I/O parity (mirror `apply_image_xarray`, `open_multi_scene`)
- [ ] MkDocs page: derivation → estimation choices → score → SNR → threshold → detection (link to the existing `00_mf_derivation.md`)
- [ ] Cross-link with `gz.plume` (#13) — that module now consumes MF outputs

## Testing

- [ ] **Bit-identity vs reference**: `MatchedFilter` on a fixed `(cube, μ, Σ_op, t)` matches `matched_filter.core.apply_image` element-wise to machine precision
- [ ] **Synthetic plume**: on `x = μ + α·t + ε` with `ε ∼ N(0, Σ)`, MF score recovers `α` in expectation; sample variance matches `1 / (tᵀ Σ⁻¹ t)` from `matched_filter_snr`
- [ ] **Null-distribution**: on a plume-free cube, MF score has mean ≈ 0 and variance ≈ `1 / (tᵀ Σ⁻¹ t)`
- [ ] **Detection FAR**: empirical FAR on Monte-Carlo plume-free cubes ≤ requested FAR for `DetectionThreshold(false_alarm_rate=p)`
- [ ] **Low-rank vs dense**: `EstimateCovLowRank(rank=k)` matches dense path to rel-tol `1e-3` once `k` reaches effective rank
- [ ] **SNR formula**: `MatchedFilterSNR(α)` equals `α · √(tᵀ Σ⁻¹ t)` to float tolerance
- [ ] **Validation**: `ValidateMFInputs` raises on zero target, raises on singular covariance, passes on valid `(Σ, t)`
- [ ] **Linear vs nonlinear target**: for small `vmr_background`, `LinearTargetFromObs` and `NonlinearTargetFromObs(amplitude=1.0)` agree within Beer–Lambert curvature
- [ ] **Welford correctness**: `StreamingBackground` over chunks `(A, B)` produces the same `(μ, Σ)` as `EstimateCovEmpirical` on `concat(A, B)` to float tolerance
- [ ] **GMM reproducibility**: identical output for fixed `random_state`
- [ ] **ApplyClusterMF**: on a two-cluster synthetic scene with known cluster-local plumes, recovers the per-cluster amplitudes within their per-cluster σ
- [ ] **Geometadata round-trip**: every Operator returning a `GeoTensor` preserves CRS, transform, and shape from the input cube
- [ ] **JIT safety**: `apply_image` runs under `jax.jit` without re-tracing on identical shapes; `validate_mf_inputs` is a no-op under tracing (already documented behaviour)
- [ ] **Performance**: 1000×1000×285 EMIT cube completes in < 5 s on CPU; low-rank path is ≥ 3× faster than dense at rank 30 (smoke benchmark in `tests/perf/`)
- [ ] **Golden fixture**: a small (128×128×40) cube with a known injected plume reproduces a reference score map within rel-tol `1e-5`

## References

- **Reference implementation**: `jejjohnson/research_notebook/projects/plume_simulation/src/plume_simulation/matched_filter/` — `__init__.py`, `core.py`, `background.py`, `cluster.py`, `streaming.py`, `target.py`, `io.py`
- Derivation + 3D-Var/PSAS one-CG connection: `notebooks/matched_filter/00_mf_derivation.md` in the same repo
- Foote et al. (2020), mag1c — sparsity-regularised methane retrieval
- Thompson et al. (2015, 2016) for AVIRIS / EMIT MF foundations
- Ledoit & Wolf (2004); Chen, Wiesel, Eldar & Hero (2010) for shrinkage / OAS
- Halko, Martinsson & Tropp (2011) for the randomised SVD used in `estimate_cov_lowrank`
- Welford (1962), Chan, Golub & LeVeque (1979) for streaming covariance
- `gaussx` — structured covariance operators (`LowRankUpdate`, `solve`), Woodbury dispatch
- `plumax` — the broader plume simulation / retrieval / emission estimation pipeline
- `lineax` — `AbstractLinearOperator`, `MatrixLinearOperator`, `DiagonalLinearOperator`
